Feat: Run in a serverful prod environment

gmarcos87 opened the pull request:
The changes made to this pull-request request allow to run the opencollective bot in a "serverful" environment.
It is currently under development and I use Heroku as a test environment.

Related https://github.com/opencollective/opencollective/issues/2603

gmarcos87 committed:
Update probot to 9.6.4

gmarcos87 committed:
build: compile typescript to javascript

gmarcos87 committed:
build: add skipLibCheck in tsconfig.json

Without this option typescript cannot compile as some dependencies do not have the
corresponding typing.

gmarcos87 committed:
build: add start method

The start method executes the code compiled with typescript.

znarf commented on package.json:
```diff
@@ -7,6 +7,9 @@
     "coverage": "yarn codecov",
+    "build": "tsc -p tsconfig.json",
+    "start": "node -r dotenv/config dist/index.js",
+    "postinstall": "npm run build",
```
heroku is now building projects automatically if I'm not mistaken https://devcenter.heroku.com/changelog-items/1557

gmarcos87 replied:
Great, I'm removing that commit to clean up the history. I wasn't aware of that change in Heroku's policy.

znarf commented on package.json:
```diff
@@ -20,7 +23,7 @@
     "multilines": "1.0.2",
     "node-fetch": "2.6.0",
-    "probot": "9.2.20"
+    "probot": "9.6.4"
```
`yarn.lock` should also be updated then

gmarcos87 replied:
done

gmarcos87 committed:
test: change contextify function

The changes incorporated in the recent version of probot make the tests fail.
The problem was how the probot.Context configuration change was simulated in the tests.
To solve this problem, the probot.Context class is extended in MockContext and the construction and configuration methods are modified.

gmarcos87 committed:
doc: add instructions to deploy in heroku

znarf commented on package.json:
```diff
@@ -7,6 +7,8 @@
     "test": "NODE_ENV=test jest",
     "coverage": "yarn codecov",
+    "build": "tsc -p tsconfig.json",
+    "start": "node -r dotenv/config dist/index.js",
```
We should not use dotenv in production environment. We expect the environment variables to be already there.

gmarcos87 replied:
done

znarf commented on tsconfig.json:
```diff
@@ -9,7 +9,8 @@
     "strict": true,
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "skipLibCheck": true
```
I guess it's a good short term solution but it would be better to not have it.

How do you explain the app is currently working fine without this flag, in development or now?

gmarcos87 replied:
I added some missing types. Now at compilation only complains of @types/promise-events that does not exist at the moment. That's why the flag.
There is an issue in probot where they end up adding the same flag to tsconfig.json
https://github.com/probot/probot/issues/512#issuecomment-393606190

I can try to find another workaround.

gmarcos87 replied:
And https://github.com/probot/probot/pull/546#issuecomment-392963951

znarf replied:
Ok, interesting.

gmarcos87 replied:
I opened a pull-request at [DefinitelyTyped repos](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/40359) to create the **@types/promise-events** package with the bare minimum. But I think it can take a long time and even never be approved. 
The fastest option would be to create the ```src/@types/promise-events``` directory with the events.d.ts file inside. This is the workaround that probot ended up implementing in its source code. 
I tried this solution, it allows us to skip the flag ```"skipLibCheck": true```

gmarcos87 committed:
build: add missing types

Add @types that were uninstalled.
The @types/promise-events has yet to be resolved.

gmarcos87 committed:
build: add custom promise-events type

Until there is a definition of type in the promise-events package or @types/promise-events is published this is the workaround that allows us to disable skipLibCheck in typescript.

gmarcos87 committed:
perf: change custom bootstrap

Change custom bootstrap in index.ts to use a more regular probot setup

gmarcos87 commented:
OK @znarf, I think it's ready for review.
I changed almost the whole content of index.ts and the way the validation paths are added. This way the bootstrap is closer to the probot standard.

znarf commented on docs/deployment.md:
```diff
@@ -27,10 +25,47 @@ Make sure that the following secrets are set (uses [now secrets](https://zeit.co
+
+In the particular case of the certificate you can use a command similar to this one to send the string to Heroku
+
+`heroku config:set PRIVATE_KEY="$(cat private-key.pem)"`
```
Doesn't look base64 encoded with that command.

znarf replied:
@gmarcos87 it doesn't seem my feedback on base64 encoding was integrated.

gmarcos87 replied:
It is now integrated in https://github.com/opencollective/opencollective-bot/pull/237/commits/d5234343aaa5dcada67065e25a00d8c46f149b64

znarf commented on docs/deployment.md:
```diff
@@ -27,10 +25,47 @@ Make sure that the following secrets are set (uses [now secrets](https://zeit.co
+| `APP_ID`            | GitHub app id                              |
+| `WEBHOOK_SECRET`    | GitHub webhook-secret                      |
+| `PRIVATE_KEY`       | One of GitHub private keys, base64 encoded |
+| `WEBHOOK_PROXY_URL` | Sentry DSN for error reporting             |
```
Mismatch, Webhook vs Sentry.

znarf reviewed (changes requested):
Looks like awesome progress! Made comments for a few minor changes.

But my major question is the following. In this state, is it still working with now v2? As we originally discussed in https://github.com/opencollective/opencollective/issues/2603#issuecomment-552591328 ? I looked quickly at branch deployment https://zeit.co/opencollective/opencollective-bot/h7euaw63v and it's outputting errors.

gmarcos87 committed:
fix: remove duplicated applications

Sentry, stats and default applications are already loaded by Probot

gmarcos87 committed:
chore: run probot as lambda in now 2.0

Wrapper around opencolective-bot to run as Lambdas in Zeit's Now 2.0

gmarcos87 commented:
Okay, it wasn't that simple to discover the solution. But it's finally working in Ziet Now v2.0 and Heroku. I removed the duplicate applications in index.ts and set Probot to run as lambda aws in Now.

znarf commented on package.json:
```diff
@@ -20,17 +22,22 @@
     "node-fetch": "2.6.0",
-    "probot": "9.2.20"
+    "probot": "9.6.4",
+    "probot-serverless-now": "^2.1.1"
```
In this project, we should use fixed dependencies everywhere.

gmarcos87 replied:
I should have guessed when I saw the rest of the dependencies were fixed. Fixed all in https://github.com/opencollective/opencollective-bot/pull/237/commits/33b595fab1b303984b825667833a6a42041fdf70

znarf reviewed (changes requested):
Looks like a fantastic job, overall a lot of simplifications. A few minor changes asked. And can could you rebase from master?

I'll have a functional try later today.

gmarcos87 committed:
doc: private_key in base64

gmarcos87 committed:
build: fixed dependencies

gmarcos87 commented:
Hi @znarf, I made the requested changes and rebase to master. It's very satisfying when you get to remove more code than you add. It doesn't happen very often.

znarf commented:
@gmarcos87 Were you able to test it? In my two attempts (one on staging, on on local environment), the bot doesn't seem to receive and process messages.

gmarcos87 commented:
> @gmarcos87 Were you able to test it? In my two attempts (one on staging, on on local environment), the bot doesn't seem to receive and process messages.

I think I understand what's going on. You are running the bot with ```npm run dev```?

znarf commented:
@gmarcos87 in one case, yes. In the other, it was on our `now` staging deployment.

gmarcos87 commented:
Reading [this issue](https://github.com/probot/create-probot-app/issues/35) in the Probot repo I see that they are not going to give support for ```npm run dev``` in ts that does not mean rebuilding again.
I updated the package.json. Now nodemon does the build and then runs ```npm run start``` over and over again.

The problem you're telling me sounds like an error in the environment variables. Can you try again? In my local execution I see arriving and leaving messages, even publishing the pull-request https://github.com/gmarcos87/lime-app/pulls

znarf commented:
@gmarcos87 hmm, there is not much background information on the issue https://github.com/probot/create-probot-app/issues/35 and I'm not sure we were facing it. Willing to try but it looked better before.

znarf commented:
Ok, it worked locally in my last attempt! I'll now check the staging configuration and try again.

gmarcos87 commented:
Another possible solution is to create a **dev-server.ts** file that runs probot with ```Probot.run(main)``` and run it from **npm run dev** (```nodemon --exec dotenv ts-node ./dev-server.ts```).

```typescript
import { Probot } from 'probot'
import main from './src/index'

Probot.run(main)
```

That would leave the operation quite similar to that which existed before this pull-request and avoids constant rebuild.

gmarcos87 committed:
fix: add dev mode for probot in typescript

znarf reviewed (approved):
Excellent, it works!

znarf commented:
Merged in https://github.com/opencollective/opencollective-bot/pull/250

Betree commented:
Hi @gmarcos87

Betree commented:
Hi @gmarcos87, we are doing a monthly Twitter thread to thank contributors. Would you be ok to be mentioned? If so, could you share your Twitter profile with me? Thanks!

gmarcos87 commented:
> Hi @gmarcos87, we are doing a monthly Twitter thread to thank contributors. Would you be ok to be mentioned? If so, could you share your Twitter profile with me? Thanks!

Hi @Betree, That would be very nice. My Twitter user is @jsPixxel
